components/menu/menu.tsx

```diff
@@ -0,0 +1,264 @@
+import Tooltip from 'components/tooltip';
+import OverlayTrigger from 'components/overlay_trigger';
+import GenericModal from 'components/generic_modal';
+
+import {MuiMenuStyled} from './menu_styled';
+
+const OVERLAY_TIME_DELAY = 500;
+const MENU_OPEN_ANIMATION_DURATION = 150;
+const MENU_CLOSE_ANIMATION_DURATION = 100;
+
+type MenuButtonProps = {
+    id: string;
+    'aria-label'?: string;
+    class?: string;
+    children: ReactNode;
+}
+
+type MenuButtonTooltipProps = {
+    id: string;
+    placement?: 'top' | 'bottom' | 'left' | 'right';
+    class?: string;
+    text: string;
+}
+
+type MenuProps = {
+    id: string;
+    'aria-label'?: string;
+
+    /**
+     * @warning Make the styling of your components such a way that they dont need this handler
+     */
+    onToggle?: (isOpen: boolean) => void;
+}
+
+interface Props {
+    menuButton: MenuButtonProps;
+    menuButtonTooltip?: MenuButtonTooltipProps;
+    menu: MenuProps;
+    children: ReactNode[];
+}
+
+/**
+ * @example
+ * import * as Menu from 'components/menu';
+ *
+ * <Menu.Container>
+ *  <Menu.Item>
+ *  <Menu.Item>
+ *  <Menu.Separator/>
+ * </Menu.Item>
+ */
+export function Menu(props: Props) {
+    const theme = useSelector(getTheme);
+
+    const isMobileView = useSelector(getIsMobileView);
+
+    const dispatch = useDispatch();
+
+    const [anchorElement, setAnchorElement] = useState<null | HTMLElement>(null);
+    const [disableAutoFocusItem, setDisableAutoFocusItem] = useState(false);
+    const isMenuOpen = Boolean(anchorElement);
+
+    function handleMenuClose(event: MouseEvent<HTMLDivElement>) {
+        event.preventDefault();
+        setAnchorElement(null);
+        setDisableAutoFocusItem(false);
+    }
+
+    function handleMenuModalClose(modalId: MenuProps['id']) {
+        dispatch(closeModal(modalId));
+        setAnchorElement(null);
+    }
+
+    function handleMenuClick() {
+        setAnchorElement(null);
+    }
+
+    function handleMenuKeyDown(event: KeyboardEvent<HTMLDivElement>) {
+        if (event.key === 'Enter') {
+            const target = event.target as HTMLElement;
+            const ariaHasPopup = target?.getAttribute('aria-haspopup') === 'true';
+
+            // Avoid closing the sub menu item on enter
+            if (!ariaHasPopup) {
+                setAnchorElement(null);
+            }
+        }
+    }
+
+    function handleMenuButtonClick(event: SyntheticEvent<HTMLButtonElement>) {
+        event.preventDefault();
+
+        if (isMobileView) {
+            dispatch(
+                openModal<MenuModalProps>({
+                    modalId: props.menu.id,
+                    dialogType: MenuModal,
+                    dialogProps: {
+                        menuButtonId: props.menuButton.id,
+                        menuId: props.menu.id,
+                        menuAriaLabel: props.menu?.['aria-label'] ?? '',
+                        onModalClose: handleMenuModalClose,
+                        children: props.children,
```

Comment:
@M-ZubairAhmed I'm seeing serialization warnings and crashes when dispatching this and redux devtools are attached.

<img width="965" alt="CleanShot 2023-01-25 at 11 22 55@2x" src="https://user-images.githubusercontent.com/11724372/214656703-5bf54b55-ee77-478b-b3dc-ad67f33ce6b9.png">
<img width="204" alt="CleanShot 2023-01-25 at 11 23 05@2x" src="https://user-images.githubusercontent.com/11724372/214656704-fcbd03fb-97ed-484f-9c50-452bd18f5997.png">

<img width="400" alt="image" src="https://user-images.githubusercontent.com/11724372/214669457-fccc5444-737f-494a-b174-c4d62b0ac024.gif">


Is there a way to avoid this? I know it'd be ideal to use `openModal`/ModalController, but if mounting it manually would avoid the mem spike in redux state, maybe it would be worth it to avoid.

Reply:
As discussed this is an edge case where this will happen when redux dev tools are open in mobile web view. Opened a ticket https://mattermost.atlassian.net/browse/MM-50006